Assets/MRTK/Core/Utilities/WebRequestRest/Rest.cs

```diff
@@ -254,31 +254,35 @@ private static async Task<Response> ProcessRequestAsync(UnityWebRequest webReque
             webRequest.disposeCertificateHandlerOnDispose = disposeCertificateHandlerOnDispose;
             await webRequest.SendWebRequest();
 
+            long responseCode = webRequest.responseCode;
+            byte[] downloadHandlerData = webRequest.downloadHandler?.data;
+            string downloadHandlerText = await Task
+                .Run(() => System.Text.Encoding.Default.GetString(downloadHandlerData));
+
 #if UNITY_2020_1_OR_NEWER
             if (webRequest.result == UnityWebRequest.Result.ConnectionError || webRequest.result == UnityWebRequest.Result.ProtocolError)
 #else
             if (webRequest.isNetworkError || webRequest.isHttpError)
 #endif // UNITY_2020_1_OR_NEWER
             {
-                if (webRequest.responseCode == 401) { return new Response(false, "Invalid Credentials", null, webRequest.responseCode); }
+                if (webRequest.responseCode == 401) { return new Response(false, "Invalid Credentials", null, responseCode); }
 
                 if (webRequest.GetResponseHeaders() == null)
                 {
-                    return new Response(false, "Device Unavailable", null, webRequest.responseCode);
+                    return new Response(false, "Device Unavailable", null, responseCode);
                 }
 
                 string responseHeaders = webRequest.GetResponseHeaders().Aggregate(string.Empty, (current, header) => $"\n{header.Key}: {header.Value}");
-                string downloadHandlerText = webRequest.downloadHandler?.text;
                 Debug.LogError($"REST Error: {webRequest.responseCode}\n{downloadHandlerText}{responseHeaders}");
-                return new Response(false, $"{responseHeaders}\n{downloadHandlerText}", webRequest.downloadHandler?.data, webRequest.responseCode);
+                return new Response(false, $"{responseHeaders}\n{downloadHandlerText}", downloadHandlerData, responseCode);
             }
             if (readResponseData)
             {
-                return new Response(true, webRequest.downloadHandler?.text, webRequest.downloadHandler?.data, webRequest.responseCode);
+                return new Response(true, downloadHandlerText, downloadHandlerData, responseCode);
             }
             else // This option can be used only if action will be triggered in the same scope as the webrequest
             {
-                return new Response(true, () => webRequest.downloadHandler?.text, () => webRequest.downloadHandler?.data, webRequest.responseCode);
+                return new Response(true, () => downloadHandlerText, () => downloadHandlerData, responseCode);
```

Comment:
These two lines are now basically identical since we're awaiting the text response earlier now. The `else` case was specifically added in https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7527 to account for an app not wanting to automatically load the text due to any CPU or garbage generation cost.

Is there any way to support this fix in that pattern? Could the `await Task.Run` call be refactored out into a method we call here if `readResponseData` is false, instead of automatically loading?

Reply:
I changed it. The only thing: when loading a large amount of text and readResponseData is false, followed by a call to ResponseBody, in my case, there is also a delay.
But if readResponseData is positive, the delay is completely eliminated.
Perhaps in the future we should make the string responseBody in the struct Response as Task<string>